make multihash pluggable

samli88 opened the pull request:
This adds the ability to register hash funcs independently and not required them in this library.

Note: this is a suggestion/WIP which can be refactored as needed.

This adds a map var which maps codecs to hash functions. The hash funcs must
have signature ~which accepts one byte slice, and returns one byte slice~ : `func(data []byte, length int) ([]byte, error)`.

It adds a RegisterHashFunc method which allows registration for any valid
codec. It prevents double registration or overwriting of hash funcs already
registered.

It also creates two funcs which are called by init() to register the funcs
which is already the current behaviour of the code:

  * registerStdlibHashFuncs
  * registerNonStdlibHashFuncs

If desired, the non-stdlib funcs can then easily be removed. Or it can be
implemented as-is, and required that future hash funcs use the Registration
func. Or refactored in some other way if necessary.

I have tested this and it is working with IPLD Dash resolver. Can also add
example in the README if this is approved / agreed.

Note: this is based upon 'cleanup' PR, #87 and needs to wait on that to be merged before this one. I can rebase after merged.

Fixes #78.

samli88 commented:
Rebased onto latest master since #87 was merged.

Stebalien commented:
Sorry for the delay, we've all been on a synchronous work sprint.

This looks great!

> The hash funcs must have signature which accepts one byte slice, and returns one byte slice.

We should *probably* allow them to return errors, even though they don't in practice. Some hash functions may have restrictions on inputs. We should probably also pass in the length (as we do for identity) as there may exist other hash functions that will simply refuse to work when the length is too short. That is: `func(data []byte, length int) ([]byte, error)`.

To avoid writing a ton of wrapper functions, I'd just modify the signatures of all the `sum*` functions.

samli88 commented:
@Stebalien Thanks, I also have been busy.

I agree that allowing for error is a good idea. I have changed the required signature as you suggested, and also refactored blake func wrappers so they work with this change, and removed panics from one (they now return errors). Now all funcs are able to be registered in init method.

Please see individual commits for more info.

Stebalien commented on sum.go:
```diff
@@ -103,17 +70,24 @@ func isBlake2b(code uint64) bool {
+	hasher, err := blake2b.New(&blake2b.Config{Size: uint8(size)})
 	if err != nil {
-		panic(err)
+		return []byte{}, err
```
This should probably return `nil`, not `[]byte{}`.

samli88 replied:
Agree, thank you!

Stebalien commented on sum.go:
```diff
@@ -103,17 +70,24 @@ func isBlake2b(code uint64) bool {
 
 	if _, err := hasher.Write(data); err != nil {
-		panic(err)
+		return []byte{}, err
```
ditto

Stebalien commented on sum.go:
```diff
@@ -125,83 +99,144 @@ func sumID(data []byte, length int) ([]byte, error) {
+}
+
+// RegisterHashFunc adds an entry to the package-level code -> hash func map.
+func RegisterHashFunc(code uint64, hashFunc func([]byte, int) ([]byte, error)) error {
```
We should document the truncation behavior:

> The hash function must return *at least* the requested number of bytes. If it returns more, the hash will be truncated.

Stebalien commented on sum.go:
```diff
@@ -36,57 +39,21 @@ func Sum(data []byte, code uint64, length int) (Multihash, error) {
+		return m, ErrSumNotSupported
+	}
+
+	var size int
```
So, we actually need to use `length` here. That is, we need to tell the hash function "we're going to truncate to size X" so it can return an error if that would be nonsensical. Unfortunately, I think we'll either still need to special case blake32, create a new function per size (probably the easiest), or pass in the size as an argument.

Basically, we need to handle the case where we request a 512bit blake2b hash and then truncate it to 256 bits.

samli88 replied:
I do not understand -- logic is the same here, right? Or is this some new logic you are suggesting for blake?

Currently, nothing is truncated from output of `blake2s.Sum256(data)`, or any blake2b from what I see. How is a special case needed? I think I cannot see your case yet. (Or you just prefer the name `length` instead of `size` for blake funcs?)

Stebalien replied:
You're right, the issue is actually with the identity function, not blake. There are two sizes here:

1. The blake "size". Really, this isn't a parameter, we just have one code per size.
2. The hash *length*. This is the length we need to pass to the hash function.

With the current code, `sumID` will *never* fail because we pass in `size` (`len(data)`), not `length`. Basically, we need to pass `length` so functions know how many bytes of the hash we're going to use.

So, my proposal is to get rid of all special casing in `Sum`. Instead, just call `funcTable[code](data, length)`.

However, we do *somehow* have to plumb the `size` through to the blake hash functions. Above, I proposed just creating a separate function per variant (using closures). We could *also* change the hash function signature to `func(data []byte, code, length int) ([]byte, error)` and let the blake2 function extract the size from the code.

samli88 replied:
Ah, I think I understand now.

> We could also change the hash function signature to func(data []byte, code, length int) ([]byte, error)

This seems ugly to me because this code is _only_ used for blake.

Also, the more I look at it, the more I think this is just a matter of names except for the ID case, where I got the logic wrong.

If I change blake funcs to use closures w/olen, it is still the same thing getting passed in w/just a different name, but these funcs (since before I touched them) never use both "length" and "size" together, only `olen`, which is equivalent to "size" for other funcs, with the exception of ID.

If I push a special case for `ID`, this should correct that logic, right? Any other case where both length and size are needed in the func would be introducing new logic which does not yet exist.

Stebalien replied:
> If I push a special case for ID, this should correct that logic, right?

It corrects the logic but the parameter still ends up meaning different things in different cases. As this is going to be an external API, it really needs to be as consistent as possible. We need to be able to unambiguously define the API.

crackcomm commented:
In hopes it can be useful here is some package I implemented lately: https://github.com/ipfn/ipfn/blob/master/pkg/digest/digest.go#L47

It does not deal with hashes other than 32 bytes but I think you could try copying:

```Go
// SumBytes - Sums hash digest using provided hasher.
func SumBytes(h hash.Hash, data ...[]byte) (digest []byte) {
	h.Reset()
	for _, body := range data {
		h.Write(body)
	}
	return h.Sum(nil)
}
```

It could remove a lot of boilerplate code.

Stebalien commented:
Status: Blocked on https://github.com/multiformats/go-multihash/pull/88#discussion_r234832655

samli88 commented:
Yes, thank you. Now I see what you mean by closures for the blake funcs. Thanks for helping w/this.

I cannot give an approval through the GitHub UI because I opened the Pull Request, but yes, I reviewed and approve these changes. 👍
